### ISSUE TITLE
Does this really need to be a module of its own?

### ISSUE THREAD

**Issue by JonRoma** [techservicesillinois/terraform-aws-ecs-shibd-data-sealer-role#4]
It would seem that this role could easily slide into `techservicesillinois/terraform-aws-iam` instead of being standalone.

**Comment by ddriddle** [comment]
It is standalone because it was intended to be shared with other Shibboleth users. It is open source code. `techservicesillinois/terraform-aws-iam` is not.

**Comment by kwessel** [comment]
That certainly made sense at the time. Now that we've retired Shib in a Box and there are a number of viable alternatives, it probably makes sense to convert it into a more general purpose module in terraform-aws-iam to create a role for ECS containers to access secret values.

**Comment by ddriddle** [comment]
I recommend archiving the repo if this is no longer in use by SDG, since it is a public repo, particularly if it is part of shib-in-a-box since it is depreciated.

**Comment by kwessel** [comment]
Yes, archiving, and not deleting, is a very good idea. I don't think Jon and I ever got to talking about what to do after merging in a more general purpose module into the terraform-aws-iam repo, but I never envisioned just deleting it. Thanks for giving us a good reason not to delete it.

**Comment by JonRoma** [comment]
Digging more deeply into IAM-related Terraform code, I find that the plot thickens. 

* I found yet another standalone repository in [terraform-aws-sm-secrets-ro-policy](https://github.com/techservicesillinois/terraform-aws-sm-secrets-ro-policy). This is called from [terraform-aws-ecs-shibd-data-sealer-role](https://github.com/techservicesillinois/terraform-aws-ecs-shibd-data-sealer-role/), which is the code that @kwessel and I wish to integrate into a submodule under [terraform-aws-iam](https://github.com/techservicesillinois/terraform-aws-iam).
* There is code in [terraform-aws-iam//roles/ecs-task-execution-role](https://github.com/techservicesillinois/terraform-aws-iam/tree/v3.2.12/roles/ecs-task-execution-role) that *appears* as if it ought to grant the GetSecretValue action to the `ecs-task-execution-role`, but which appeared to not work until we added the `shibd-data-sealer` role using from [terraform-aws-ecs-shibd-data-sealer-role](https://github.com/techservicesillinois/terraform-aws-ecs-shibd-data-sealer-role/).
* In addition, I see code in the `ecs-task-execution-role` that appears to give the grantee access to **all** SSM parameters and **all** SecretsManager secrets. Yikes! This was OK for expediency when we were new to AWS (and IAM in particular), but we shouldn't give Course Explorer access to the Shibboleth IdP's secrets, for example. Keith and I have refactored the IdP secrets manager code to make a hierarchy (e.g., `shibboleth-idp/data-sealer` instead of `shibboleth-idp-data-sealer`. With this hierarchy in place, we could restrict future access to SecretsManager secrets using more precise wildcard patterns.
* I looked at [terraform-aws-iam](https://github.com/techservicesillinois/terraform-aws-iam) as a whole last evening, and I became discouraged. It's a rat's nest with a lot of technical debt and not much documentation. This is understandable given that it was code we wrote early in the game to get off the ground, but these discoveries do not make the terraform-aws-iam a good showpiece to present to campus. There's no question that Amazon IAM is complex, but its inherent complexity is also a compelling reason to give this module an in-depth review for baggage that remains from our early implementation, and to give each submodule some documentation so that people who are not Amazon IAM experts can make sense of it, and make good decisions about how to set up their access control.
* After @ddriddle left SDG, I added the [terraform-aws-iam//service-roles](https://github.com/techservicesillinois/terraform-aws-iam/tree/v3.2.12/service-role) submodule to eliminate the annoyance that I and service owners had because each of the `xxxAdmin` roles was a one-off role definition that baked in service-specific policies into modules; the new code means one can simply modify an infrastructure definition and grant access to a lot of things. Apropos of this, there are a few AWS services that need to be added to this module, but I have also reached the conclusion that I misnamed the module. Perhaps `service-user` would be more descriptive of the function than the present name.

My conclusion here is that fixing the data sealer role stuff is just the tip of the iceberg. I think this whole module deserves a line-by-line review, which will probably flush out some code changes as well as structural changes to the hierarchy.

You already know the reasons that terraform-aws-iam is not a public module, even though I would like it to be so.

I think @edthedev and @tim-morgan are also stakeholders if we intend to push this out to a wider audience. There's a lot of useful code there, but it suffers from being specific to our business needs, and from being written early in our cloud development days.

I would probably name terraform-aws-iam as one of the earliest modules we should schedule for a collaborative review. If someone wants to advocate making this public, I certainly won't object.

BTW, any time we retire a module, we ought to archive it rather than simply deleting it. How long it should remain as archived isn't important until someone asks for support for something we deprecated five years ago. :-)
